### DISCUSSION TITLE
Downloading a file from own repository

### DISCUSSION THREAD

**Discussion by RBTCGP** [community/community#23955] [New to GitHub]
<p>How can I download a file from my repository?</p>
<p>I can log in or I can request something with a token from</p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/362;"><img src="https://user-images.githubusercontent.com/81124275/181080383-55252aee-0a2b-4f6b-9e71-65ba5305ea81.png" class="thumbnail" width="690" height="362" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/d/9/d9a7f96b8e2a275c2c999d57b275d346cee251a0_2_690x362.png, https://aws1.discourse-cdn.com/github/optimized/3X/d/9/d9a7f96b8e2a275c2c999d57b275d346cee251a0_2_1035x543.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/d/9/d9a7f96b8e2a275c2c999d57b275d346cee251a0.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/d/9/d9a7f96b8e2a275c2c999d57b275d346cee251a0_2_10x10.png"></div>

<h3><a href="https://github.com/" target="_blank" rel="noopener nofollow ugc">GitHub: Where the world builds software</a></h3>

<p>GitHub is where over 56 million developers shape the future of software, together. Contribute to the open source community, manage your Git repositories, review code like a pro, track bugs and feat...</p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>But I always get 404 as response.</p>

**Comment by airtower-luna** [comment]
<p>Could you describe exactly how you’re trying to download the file?</p>

**Comment by RBTCGP** [comment]
<p>I tried it with the urls given by github and https.</p>

**Comment by RBTCGP** [comment]
<p>I do not know, how I can get a valid URL for a file in my repository to be able to download it. It seams that the URLs under “raw” with the token expire very soon.</p>
<p>And with login and RBTCGP/NeoLoad/blob/komplett/PH78.yaml I get “The connection was unexpected disconnected”</p>
<p>wget <a href="https://RBTCGP:PASSWORD@github.com/RBTCGP/NeoLoad/blob/komplett/PH78.yaml" rel="noopener nofollow ugc">https://RBTCGP:PASSWORD@github.com/RBTCGP/NeoLoad/blob/komplett/PH78.yaml</a>        wget : Die Anfrage wurde abgebrochen: Die Verbindung wurde unerwartet getrennt…<br>
In Zeile:1 Zeichen:1</p>
<ul>
<li>wget <a href="https://RBTCGP:PASSWORD@github.com/RBTCGP/NeoLoad/blob/komplett/" rel="noopener nofollow ugc">https://RBTCGP:PASSWORD@github.com/RBTCGP/NeoLoad/blob/komplett/</a> …</li>
<li>
<pre><code class="lang-auto">  + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], WebExc
 eption
  + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand</code></pre>
</li>
</ul>

**Comment by airtower-luna** [comment, marked as answer]
<p>You’ll need to create a PAT and send that in the <code>Authorization</code> header to download from the raw URL, as described here:</p>
<aside class="quote" data-post="8" data-topic="168635">
  <div class="title">
    <div class="quote-controls"></div>
    
    <a href="https://github.community/t/how-to-get-token-of-a-raw-github-file/168635/8">How to get token of a raw GitHub file</a> <a class="badge-wrapper  bullet" href="/c/github-help/how-to-use-git-and-github/21"><span class="badge-category-parent-bg" style="background-color: #6a737d;"></span><span class="badge-category-bg" style="background-color: #ffd1ac;"></span><span style="" data-drop-close="true" class="badge-category clear-badge" title="Ask and answer questions regarding Git and GitHub.">How to use Git and GitHub</span></a>
  </div>
  <blockquote>
    That’s not going to work, because those tokens have a limited lifetime. You’ll have to <a href="https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token">create a Personal Access Token (PAT)</a> and download the files using that, using the Authorization header instead of a token in the URL. 
Something like this: 
curl -H "Authorization: token ${PAT}" https://raw.githubusercontent.com/user/repo/main/file.txt
  </blockquote>
</aside>

<p>That example uses curl, but you should be able to do the same thing with the <code>--header</code> option of wget.</p>

**Comment by RBTCGP** [comment]
<p>Many thanks, it works!</p>
<p>Nevertheless, is there also a possibility to get a link to a zip of the files in a repository branch without putting them into a folder and zipping the folder? I need the files in the root of the folder.</p>

**Comment by airtower-luna** [comment]
<p>You can use the URL behind the “Download ZIP” field in the popup that shows up if you click the green “Code” button on your repository page. The <code>Authorization</code> header works just the same. :slightly_smiling_face:</p>

**Comment by RBTCGP** [comment]
<p>Yes, but this puts the files into a folder and zips the folder. I need the files in the root of the zipped folder without any folder inbetween.</p>

**Comment by airtower-luna** [comment]
<p>There’s no built-in mechanism for that. You could set up an Actions workflow to build a ZIP just like you need it, but I’d just move files into place after extraction (write a script to do that) unless the ZIP needs to be processed by some tool you can’t adjust.</p>
